### PR TITLE
Riyo - Sockets

### PR DESCRIPTION
**Pull request by RPerry**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | You are notified that an incorrect argument has been attempted, such as the wrong number of parameters when calling a method which could cause the program to break.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | So that the methods can be called without having to create a particular instance of order or customer.
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | It is a many-to-many relationship as there is one customer to many orders orders. It is different in that solar system was clearly encompassing of the planets in a one to many relationship but there are many orders and many customers, also who have multiple orders.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | The customer is tracked within order through the customer id. In the program it is tracked also through the customer id in the form of the Customer instance for each order.
Did the presence of automated tests change the way you thought about the problem? How? | It slightly changed how I went about the problem by ensuring that edge cases that I would not have thought of until it failed were considered in the initial design.
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? | Devin said we could skip this question

### CODE REVIEW

File: lib/customer.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    customer_data_array.each do |customer|
+      customer[2] = customer[2] + " " + customer[3] + " " + customer[4] + " " + customer[5]
+      customer.slice!(3..5)
+    end
+
+    customer_data_array.each do |customer|
+      customer_data_collection << Customer.new(customer[0].to_i, customer[1], customer[2])
+    end
```
Comment (dHelmgren) [inline lib/customer.rb:28]: Is there a way that you could combine these three loops?

File: lib/customer.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  def initialize(id, email, address)
+    @id = id
+    @email = email
+    @address = address
+  end
+
+  def self.all
+    customer_data_array = []
+    customer_data_collection = []
+    CSV.open("../data/customers.csv", "r").each do |row|
```
Comment (dHelmgren) [inline lib/customer.rb:17]: Quick note on path conventions! We assume that we are calling these files from the root of the directory, not from lib. This will make more sense once we get to rails, but just keep that in mind: when you run rake, run it from the root of the project. The path we recommend is `data/customers.csv`

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    end
+    return product_hash
+  end
+
+  def self.all
+    order_collection = []
+    CSV.open("../data/orders.csv", "r").each do |row|
+      id = row[0].to_i
+      products = Order.make_hash(row[1])
+      customer_id = Customer.find(row[2].to_i)
```
Comment (dHelmgren) [inline lib/order.rb:58]: Good work getting this tricky order-parsing logic figured out!

**Comment by dHelmgren** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | I would argue that this is a one-to-many relation. While there's certainly only one customer for each order, there might be many orders for a single customer; in other words, multiple orders in the CSV file have the same customer ID.<br><br>In fact, your code already reflects this. Your implementation of `Order.find_by_customer` returns a _list_ of orders, not just one order, which implies that there might be many orders for one customer. For the Big O, here's the answer: Your implementation of `Order.find` calls `Order.all`, which in turn calls `Customer.find` once for each order. That means that if the complexity of `Customer.find` is `O(C)`, then `Order.find` is `O(C) * O(R) = O(C*R)`.
Used Git Regularly | I know you hadn't seen my last set of comments yet, but you should commit more! A project this size should probably have 5-10 commits.
**Wave 1** |
All provided tests pass | Yes
Using the appropriate `attr_` for instance variables | yes
**Wave 2** |
All stubbed tests are implemented fully and pass | yes
Used `CSV` library only in `.all` (not in `.find`) | yes
Appropriately parses the product data from CSV file in `Order.all` | yes
`Order.all` calls `Customer.find` to set up the composition relation | yes
**Additional Notes** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review, but it's clear that the learning goals for this assignment were met. Keep up the hard work!
